traitlets/config/loader.py

```diff
@@ -628,15 +698,76 @@ def _load_flag(self, cfg):
 
 flag_pattern = re.compile(r'\-\-?\w+[\-\w]*$')
 
+
+class _KVAction(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        values = ["-" if v is _DASH_REPLACEMENT else v for v in values]
+        items = getattr(namespace, self.dest, None)
+        if items is None:
+            items = DeferredConfigList()
+        else:
+            items = DeferredConfigList(items)
+        items.extend(values)
+        setattr(namespace, self.dest, items)
+
+
+_DOT_REPLACEMENT = "__DOT__"
+_DASH_REPLACEMENT = "__DASH__"
+
+class _DefaultOptionDict(dict):
+    """Like the default options dict
+
+    but acts as if all --Class.trait options are predefined
+    """
+
+    def _add_kv_action(self, key):
+        self[key] = _KVAction(
+            option_strings=[key],
+            dest=key.lstrip("-").replace(".", _DOT_REPLACEMENT),
+            nargs="+",
+        )
```

Comment:
So this is the culprit; and I think this approach is 100% incompatible with trailing varargs.

Reply:
I thought we needed `--foo` to be specified multiple times, i.e. `--foo a=1 --foo b=2`. That was the goal, at least. I agree and fear it's going to be too ambiguous if that's not the case. I'll look into the variable-arguments docs in argparse to see if we can exert more control over this.